### PR TITLE
ckb-debugger support decode instruction

### PR DESCRIPTION
**Pull request by joii2020**

### CODE REVIEW

File: ckb-debugger/src/decode.rs [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            OP_OR => Rtype(self.instruction).to_string(),
+            OP_AND => Rtype(self.instruction).to_string(),
+            OP_SLL => Rtype(self.instruction).to_string(),
+            OP_SLLW => Rtype(self.instruction).to_string(),
+            OP_SRL => Rtype(self.instruction).to_string(),
+            OP_SRLW => Rtype(self.instruction).to_string(),
+            OP_SRA => Rtype(self.instruction).to_string(),
+            OP_SRAW => Rtype(self.instruction).to_string(),
+            OP_SLT => Rtype(self.instruction).to_string(),
+            OP_SLTU => Rtype(self.instruction).to_string(),
```
Comment (XuJiandong) [inline ckb-debugger/src/decode.rs]: These match arms can be merged.

File: ckb-debugger/src/decode.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            isa,
+        }
+    }
+
+    pub fn get_isa_str(&self) -> &str {
+        match self.isa {
+            ISA::Imc => "Base Instruction",
+            ISA::A => "A extension",
+            ISA::B => "B extension",
+            ISA::Mop => "MOP extension",
```
Comment (XuJiandong) [inline ckb-debugger/src/decode.rs]: RISC-V doesn't have MOP extension.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Review by mohanson — CHANGES_REQUESTED** [review]
You can use TaggedInstruction to get this job done

```rs
if let Some(i) = instructions::i::factory::<u64>(instruction, VERSION2) {
          let tagged_instruction_option = instructions::tagged::TaggedInstruction::try_from(i);
          if let Ok(tagged_instruction) = tagged_instruction_option {
              println!("{}", tagged_instruction)
          } else {
              println!("unknown instruction");
          }
}
```
result:

```
addi a7,93(zero) # if instruction is 0x05d00893
```

File: ckb-debugger/src/decode.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+enum ISA {
+    Imc,
+    A,
+    B,
+    Unknow,
+}
+
+pub fn decode_instruction(data: &str) -> Result<(), Box<dyn std::error::Error>> {
+    let has_0x = data.find("0x");
```
Comment (mohanson) [inline ckb-debugger/src/decode.rs]: data.start_with("0x")

File: ckb-debugger/src/decode.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        let tagged_instruction_option = tagged::TaggedInstruction::try_from(self.instruction);
+        if let Ok(tagged_instruction) = tagged_instruction_option {
+            tagged_instruction.to_string()
+        } else {
+            panic!("unknown instruction")
+        }
+    }
+
+    fn get_isa(ins_opcode: InstructionOpcode) -> ISA {
```
Comment (mohanson) [inline ckb-debugger/src/decode.rs]: ckb-vm does not guarantee that the instructions in an instruction set are ordered together. For example branch instructions in IMC are not grouped with other IMC instructions, See: https://github.com/nervosnetwork/ckb-vm/blob/develop/definitions/src/instructions.rs#L270-L287

**Comment by mohanson** [comment]
I think there is an easier way to do it, just do things like below

```rs
if let Some(i) = instructions::i::factory::<u64>(instruction, VERSION2) {
    let tagged_instruction = instructions::tagged::TaggedInstruction::try_from(i).unwrap();
    println!("{} (ISA_I)", tagged_instruction)
}

if let Some(i) = instructions::m::factory::<u64>(instruction, VERSION2) {
    let tagged_instruction = instructions::tagged::TaggedInstruction::try_from(i).unwrap();
    println!("{} (ISA_M)", tagged_instruction)
}

if let Some(i) = instructions::a::factory::<u64>(instruction, VERSION2) {
    let tagged_instruction = instructions::tagged::TaggedInstruction::try_from(i).unwrap();
    println!("{} (ISA_A)", tagged_instruction)
}

if let Some(i) = instructions::b::factory::<u64>(instruction, VERSION2) {
    let tagged_instruction = instructions::tagged::TaggedInstruction::try_from(i).unwrap();
    println!("{} (ISA_B)", tagged_instruction)
}

if let Some(i) = instructions::rvc::factory::<u64>(instruction, VERSION2) {
    let tagged_instruction = instructions::tagged::TaggedInstruction::try_from(i).unwrap();
    println!("{} (ISA_RVC)", tagged_instruction)
}
```

**Review by mohanson — APPROVED** [review]

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**